### PR TITLE
Restricts what roles can be added to user discord profile

### PR DESCRIPTION
**Pull request by Ajeyakrishna-k**
Date: `30 November 2023`

Developer Name: @Ajeyakrishna-k 

----

## Description: 
Currently anyone can pass any discord role id and get privileges. This pr fixes that.

Is Under Feature Flag 
- [ ] Yes
- [x] No

Database changes
- [ ] Yes
- [x] No

Breaking changes (If your feature is breaking/missing something please mention pending tickets)
- [ ] Yes
- [x] No

Is Development Tested?

- [ ] Yes
- [x] No

Tested in staging?

- [ ] Yes
- [x] No

### Add relevant Screenshot below ( e.g test coverage etc. )
Screenshot 2023-11-30 at 6 39 50 AM

Screenshot 2023-11-30 at 6 38 14 AM

### CODE REVIEW

**Review by vinayak-trivedi — DISMISSED** [review]
LGTM

**Comment by heyrandhir** [comment]
Can you also add the coverage stats.

File: controllers/discordactions.js [hunk 1]
```diff
@@ -146,6 +157,20 @@ const addGroupRoleToMember = async (req, res) => {
...
+    });
+    const userDataPromise = fetchUser({ discordId: userid });
+    const [{ roleExists }, userData] = await Promise.all([roleExistsPromise, userDataPromise]);
+
+    if (!roleExists) {
+      res.boom.forbidden("Permission denied. Cannot delete the role.");
+    }
+    if (req.userData.id !== userData.user.id) {
+      res.boom.forbidden("Permission denied. Cannot delete the role.");
+    }
```
[the remark this answers is not in the document]
Comment (Ajeyakrishna-k) [reply controllers/discordactions.js:172]: done.

File: test/unit/models/discordactions.test.js [hunk 2]
```diff
@@ -87,51 +87,59 @@ describe("discordactions", function () {
...
-        existingRole.push(roleData);
-      };
-
-      getStub.returns({
-        limit: sinon.stub().resolves({
-          empty: false,
-          forEach: callbackFunction,
-        }),
+    it("should return false if roleid doesn't exist in the database", async function () {
+      const roleid = "Test Role";
```
[the remark this answers is not in the document]
Comment (Ajeyakrishna-k) [reply test/unit/models/discordactions.test.js:110]: Then actual roleid will be overwritten right? We are using it in other tests

[the remark this answers is not in the document]
Comment (heyrandhir) [reply test/unit/models/discordactions.test.js:110]: Yeah correct.

**Review by heyrandhir — APPROVED** [review]

File: test/integration/discordactions.test.js [hunk 3]
```diff
@@ -229,16 +287,34 @@ describe("Discord actions", function () {
...
+    it("should not allow unknown role to be deleted from user", async function () {
+      const res = await chai
+        .request(app)
+        .post("/discord-actions/roles")
+        .set("cookie", `${cookieName}=${jwt}`)
+        .send({ roleid: "randomId", userid: "abc" });
+
+      expect(res).to.have.status(403);
+      expect(res.body).to.be.an("object");
+      expect(res.body.message).to.equal("Permission denied. Cannot add the role.");
```
[the remark this answers is not in the document]
Comment (Ajeyakrishna-k) [reply test/integration/discordactions.test.js]: Method name was patch

[the remark this answers is not in the document]
Comment (Ajeyakrishna-k) [reply test/integration/discordactions.test.js]: changed it to delete

**Review by heyrandhir — APPROVED** [review]

**Pull request merged**